apps/emqx_rule_engine/src/emqx_rule_actions.erl

```diff
@@ -188,7 +201,7 @@ on_action_republish(Selected, _Envs = #{
             id = emqx_guid:gen(),
             qos = if TargetQoS =:= -1 -> 0; true -> TargetQoS end,
             from = ActId,
-            flags = #{dup => false, retain => false},
+            flags = #{dup => false, retain => if TargetRetain =:= 0 -> false; true -> true end},
```

Comment:
And this should also be done at the last function clause of `on_action_republish/2`, at line 180. To make sure a PUBLISH message is republished using a specified retain flag.